### ISSUE TITLE
ERROR:  tuple already updated by self

### ISSUE THREAD

**Issue by derekjn** [pipelinedb/pipeline_kafka#40]
Saw this when starting up multiple `topic` `->` `stream` consumers:

```
ERROR:  tuple already updated by self
```

**Comment by usmanm** [comment]
Vinay @vinaygulani1 19:00
@usmanm thanks. I am getting an error "ERROR: tuple already updated by self
LOG: worker process: [kafka consumer] stream <- topic (PID 4171) exited with exit code 1. "

**Comment by usmanm** [comment]
This seems like a bug introduced recently with the consumer group stuff. Hadn't seen it before.

**Comment by vinaygulani1** [comment]
Any news on this. 
@usmanm would you like to have a quick chat on this and see if I can make the changes necessary. Please let me know.

**Comment by michaelsauter** [comment]
@derekjn Any updates on this one? Seeing this pop up randomly after some time. Then it is quite hard to get rid of it. Right now, I'm calling `pipeline_kafka.consume_end`, then removing from `pipeline_kafka.consumers` and `pipeline_kafka.offsets`, and finally dropping stream and view. Only then is it possible to setup up everything again.

I only have one stream per topic, and one view per stream. I'm consuming 5-10 topics.

**Comment by michaelsauter** [comment]
I just noticed that when I look into `pipeline_kafka.offsets` for the consumers that die, the stored offsets look weird. The messages in Kafka are evenly distributed across the partitions, but the offets in PipelineDB are not. Some partitions are at the latest offsets, but others are way behind ...

**Comment by derekjn** [comment]
> @derekjn Any updates on this one?

@michaelsauter I don't believe we were ever able to repro this, and haven't seen it since but we're more than happy to take a look at it.

> Seeing this pop up randomly after some time. Then it is quite hard to get rid of it. Right now, I'm calling pipeline_kafka.consume_end, then removing from pipeline_kafka.consumers and pipeline_kafka.offsets, and finally dropping stream and view. Only then is it possible to setup up everything again.

Ugh, sorry to hear that--that's not good at all. From what I can tell, the bad behavior you're seeing looks like this:

* You see a `tuple already updated by self` error thrown by a consumer process
* The processes that throw this error essentially die
* The offsets are essentially in an unusable state after that

Please add anything I missed here or correct me if I am mistaken.

One question I have for you at this point is: does the `tuple already updated by self` appear repeatedly? Or is it just once?

**Comment by michaelsauter** [comment]
When `tuple already updated by self` appears, then the consumer dies, and it is being started again, and `tuple already updated by self` appears again .... and so on in an endless loop.

The only way to stop it is what I described above - ending consumption, deleting everything and starting again from scratch.

The continuous view I'm filling is an aggregation of the individual messages. The count column increases forever if that problem happens ...

UPDATE:
Regarding my last point about the ever-increasing counts, I started the consumer with `format := 'json', start_offset:= -2` originally, and my assumption is that the consumer reads some messages over and over again because some offsets are not committed properly.

**Comment by derekjn** [comment]
@michaelsauter got it, this is helpful. We're looking into this and will report back when we know more.

**Comment by michaelsauter** [comment]
@derekjn Any news on this one?

It hasn't occured the last week or so, but just now it happened again, right when I added a new stream/view/consumer - at that time another consumer which ran fine for a longer time failed. Then I deleted that old consumer (with stream and view), recreated it and now all are running again ...

**Comment by derekjn** [comment]
@michaelsauter nothing just yet but that's actually good information. We will stress test against that scenario and hopefully get to the bottom of this soon.

**Comment by derekjn** [comment]
@michaelsauter can you try running the latest `master` now? We've had some trouble reproducing the exact behavior you're describing but have a couple of theories as to what might be going wrong. We just pushed a small change that addresses one of these theories.

And if it does happen again, can you share with us anything at all that you see in the logs?

**Comment by michaelsauter** [comment]
@derekjn Sorry for the late reply. Thanks for looking into it!

I didn't find time yet to update to latest master, but I hope to get to it end of next week and then I'll report back here.

**Comment by michaelsauter** [comment]
@derekjn I just encountered the issue again, then build a new image and I'm running latest master now. So far no issues.

**Comment by derekjn** [comment]
> @derekjn I just encountered the issue again, then build a new image and I'm running latest master now. So far no issues.

@michaelsauter just to clarify, did you run into the issue again on an older build, and now you're running the latest `master`?

**Comment by michaelsauter** [comment]
@derekjn Sorry for the confusion. I had the issue while still on the old build. Afterwards I made a new image from latest master and am running that now - no issues so far.

**Comment by derekjn** [comment]
@michaelsauter got it, we'll stand by for any updates. Thanks!

**Comment by michaelsauter** [comment]
It was running fine the since then. Today I stopped all consumers, dropped all views and recreated them from the Kafka topics (we keep those messages a long time). Unfortunately now I have encountered `ERROR:  tuple concurrently updated`.  So not the same problem, and the effect is different (consumer exits, but doesn't go crazy), but there is still some issue ...

I wonder if it has to do with load? When I recreate the views, CPU is at 100%, memory is fine.
